File: leetcode/invert_binary_tree/invert_binary_tree.java
package invert_binary_tree;

import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public void printTopDown(TreeNode root) {
        if(root == null) return;
        Queue<TreeNode> treeNodeQueue = new LinkedList<>();
        treeNodeQueue.add(root);

        TreeNode curr;
        while (!treeNodeQueue.isEmpty()) {
            curr = treeNodeQueue.poll();

            System.out.print(curr.val + " ");
            if(curr.left != null){
                treeNodeQueue.add(curr.left);
            }

            if(curr.right != null) {
                treeNodeQueue.add(curr.right);
            }
        }
        System.out.println();
    }
}

class Solution {
    public TreeNode invertTree(TreeNode root) {
        Queue<TreeNode> treeNodeQueue = new LinkedList<>();
        treeNodeQueue.add(root);

        TreeNode curr, tmp;
        while(!treeNodeQueue.isEmpty()) {
            curr = treeNodeQueue.poll();

            tmp = curr.left;
            curr.left = curr.right;
            curr.right = tmp;

            if(curr.left != null) {
                treeNodeQueue.add(curr.left);
            }

            if(curr.right != null) {
                treeNodeQueue.add(curr.right);
            }
        }
        return root;
    }

    public static void main(String[] args) {
        //[4,2,7,1,3,6,9]
        TreeNode root = new TreeNode(
            4,
            new TreeNode(
                2,
                new TreeNode(1),
                new TreeNode(6)
            ),
            new TreeNode(
                7,
                new TreeNode(3),
                new TreeNode(9)
            )
        );

        root.printTopDown(root);
        new Solution().invertTree(root);
        root.printTopDown(root);
    }
}

File: leetcode/invert_binary_tree/invert_binary_tree.cpp
/*
 * Time Complexity: O(N)
 *   - Each node is visited once.
 *
 * Space Complexity: O(H)
 *   - H is the height of the tree (recursion stack).
 */

#include<iostream>
#include<queue>

using std::queue;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}

    void printTopDown(TreeNode* root) {
        if(root == nullptr) return;
        queue<TreeNode*> treeNodeQueue;
        treeNodeQueue.push(root);

        TreeNode *curr;
        while(!treeNodeQueue.empty()){
            curr = treeNodeQueue.front();
            treeNodeQueue.pop();

            std::cout << curr->val << " ";

            if(curr->left != nullptr) {
                treeNodeQueue.push(curr->left);
            }

            if(curr->right != nullptr) {
                treeNodeQueue.push(curr->right);
            }
        }
        std::cout << std::endl;
    }
};

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(root == nullptr) return root;
        queue<TreeNode*> treeNodeQueue;
        treeNodeQueue.push(root);

        TreeNode *curr, *tmp;
        while(!treeNodeQueue.empty()){
            curr = treeNodeQueue.front();
            treeNodeQueue.pop();

            tmp = curr->left;
            curr->left = curr->right;
            curr->right = tmp;

            if(curr->left != nullptr) {
                treeNodeQueue.push(curr->left);
            }

            if(curr->right != nullptr) {
                treeNodeQueue.push(curr->right);
            }
        }
        return root;
    }
};

int main() {
    TreeNode root (
        4,
        new TreeNode(
            2,
            new TreeNode(1),
            new TreeNode(6)
        ),
        new TreeNode(
            7,
            new TreeNode(3),
            new TreeNode(9)
        )
    );

    root.printTopDown(&root);

    Solution s;
    s.invertTree(&root);

    root.printTopDown(&root);
}